### ISSUE TITLE
DependsOn of a component should implicitly depend on all its children

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#1743]
Today, passing a `ComponentResource` as a `DependsOn` resource will not do anything.  Users typically try to do this as a way to say "I want to depend on everything in this component being created before creating this other resource".  We should make that the meaning of `dependsOn: [component]`.

Of course, this will be fairly coarse grained, and may involve depending on more than strictly necessary - so component authors and users may want to expose capabilities to depend on only certain critical pieces so that other work can proceed in parallel with part of the component creation.  But the right default meaning for this seems to be the conservative "depend on all my children".

**Comment by swgillespie** [comment]
This could be tricky. The current semantics of `RegisterResource` are that the step is ready to execute at the moment that we see the event. In order to do this we'd need to have access to a component resource's list of child resources in the language host so that we can wait for them to resolve before invoking the engine.

More palatable to me sounds like relaxing the invariant on `RegisterResource` and allowing unready requests enter the engine. The engine could still require that all properties be resolved, but it could wait until all `dependsOn` resources resolve before scheduling the step to execute. This jives with the notion of the engine being a generalized DAG executor a little more than the former does.

**Comment by hausdorff** [comment]
This bug is extremely annoying when you're trying to do something like force one Helm chart to depend on another. There is a work-around, which is to have one Helm chart depend on a resource in the other Helm chart that gets deployed last (usually a `Service`), but at the outset it can be very confusing why component resources seem to be rolling out simultaneously when they should be sequential.

**Comment by joeduffy** [comment]
I have to admit, something has always felt fundamentally broken with component resources here.

When implementing components, my original approach was to have a `BeginRegisterResource` and `EndRegisterResource` pair of functions; the component's constructor then took a callback so that we had a definite lifetime of a component's initialization (that is, begin, initialize all children, end (with outputs)). This had the added advantage that output properties were simply returned by the callback, rather than being ad-hoc registrations that happen an unknown moments in time. 

We moved away this for two reasons: (1) the engine didn't support parallelism at the time, and (2) the callback-style of initialization suffered usability problems. But now, looking backwards, we have parallelism and component authors are probably okay with a minor hit to usability, given the advantages.

The fact that we don't have this sort of structured lifetime over component initialization honestly makes every other approach feel a bit hacky to me.

**Comment by lukehoban** [comment]
We don't have a design for this yet, so pushing to M20.  @pgavlin Do you have thoughts/notes on how you expect we'll tackle this?

**Comment by lukehoban** [comment]
Related to https://github.com/pulumi/pulumi/issues/2295 which may be the way we address this.

**Comment by lukehoban** [comment]
We ended up backing out the change that "fixed" this - so reopening to track the desire to support this better in the future.

**Comment by CyrusNajmabadi** [comment]
> We ended up backing out the change that "fixed" this - so reopening to track the desire to support this better in the future.

When did that happen?  I still see the logic here:

```ts
    const allDirectDependencyURNs = await getAllTransitivelyReferencedCustomResourceURNs(explicitDirectDependencies);
...
async function getAllTransitivelyReferencedCustomResourceURNs(resources: Set<Resource>) {
    // Go through 'resources', but transitively walk through **Component** resources, collecting any
    // of their child resources.  This way, a Component acts as an aggregation really of all the
    // reachable custom resources it parents.  This walking will transitively walk through other
    // child ComponentResources, but will stop when it hits custom resources.  in other words, if we
    // had:
    //
    //              Comp1
    //              /   \
    //          Cust1   Comp2
    //                  /   \
    //              Cust2   Cust3
    //              /
    //          Cust4
    //
    // Then the transitively reachable custom resources of Comp1 will be [Cust1, Cust2, Cust3]. It
    // will *not* include `Cust4`.

    // To do this, first we just get the transitively reachable set of resources (not diving
    // into custom resources).  In the above picture, if we start with 'Comp1', this will be
    // [Comp1, Cust1, Comp2, Cust2, Cust3]
    const transitivelyReachableResources = getTransitivelyReferencedChildResourcesOfComponentResources(resources);

    const transitivelyReachableCustomResources =  [...transitivelyReachableResources].filter(r => CustomResource.isInstance(r));
    const promises = transitivelyReachableCustomResources.map(r => r.urn.promise());
    const urns = await Promise.all(promises);
    return new Set<string>(urns);
}
```

And so on.

**Comment by CyrusNajmabadi** [comment]
I don't know what's being asked of here.  AFAICT, the logic we have is the intended logic.  Effectively, depending-on a ComponentResource is the same as depending on all transitively reachable child custom-resources of it.

**Comment by CyrusNajmabadi** [comment]
I'm going to close this out.  @lukehoban  if you do think there's work here, please let me know.  specifically, can you clarify:

> We ended up backing out the change that "fixed" this - so reopening to track the desire to support this better in the future.

**Comment by nesl247** [comment]
I'm not sure if this should have been closed or not. Maybe we need to open up a ticket specific for helm? Because you cannot currently depend upon a helm chart.

**Comment by lukehoban** [comment]
Interesting - yes - there could be a Helm-specific issue here. Could you open a new issue on that in the kubernetes repo?

**Comment by nesl247** [comment]
I'm actually surprised this was closed given helm was specifically mentioned here: https://github.com/pulumi/pulumi/issues/1743#issuecomment-424606039

**Comment by lukehoban** [comment]
> I'm actually surprised this was closed given helm was specifically mentioned here: #1743 (comment)

Yeah - I think there *was* originally a much more general issue here - and that was fixed - which addressed what the title and most of the commentary here were referencing.

The fact that there was a *separate* issue specific to Helm was not immediately clear - and I do think we should track that separately as it will require a specific fix in that library.

> Edit: Ticket opened at pulumi/pulumi-kubernetes#861

Thanks!

**Comment by benlongo** [comment]
Is this only implemented for the nodejs sdk?
